src/ocean/util/container/btree/Implementation.d

```diff
@@ -0,0 +1,1121 @@
+            searched element, or ElementType.init if not found.
+
+    *******************************************************************************/
+
+    package ElementType search (ref ElementType element, out bool found)
+    {
+        size_t index;
+        auto node = this.search(element, index);
+        if (node) found = true;
+
+        return node is null? ElementType.init : node.elements[index];
+    }
+
+    /***************************************************************************
+
+        Finds the requested element in the tree.
+
+        Params:
+          element = element to search for
+          tree = tree to search in
+          index = index of the element in the node
+
+        Returns: pointer to the node holding `element`, null otherwise
+
+    ***************************************************************************/
+
+    private BTreeNode* search (ref ElementType element,
+        out size_t index)
+    {
+        /// internal function for recursion
+        /// which does the actual search (outer function just accepts the tree)
+        static BTreeNode* searchImpl (BTreeNode* root,
+            ref Const!(ElementType) element, out size_t index)
+        {
+            auto pos = 0;
+
+            // TODO: binary search
+            while (pos < root.number_of_elements
+                && element > root.elements[pos])
+            {
+                pos++;
+            }
+
+            // now pos is the least index in the key array such
+            // that f < elements[pos]
+            if (pos < root.number_of_elements
+                && element == root.elements[pos])
+            {
+                index = pos;
+                return root;
+            }
+
+            // Nowhere to descend
+            if (root.is_leaf)
+            {
+                return null;
+            }
+
+            return searchImpl(root.child_nodes[pos], element, index);
+        }
+
+        if (this.root is null)
+        {
+            return null;
+        }
+
+        return searchImpl(this.root, element, index);
+    }
+
+    /******************************************************************************
+
+        Does the necessary traversal to delete an element and rebalance parents
+        in the process.
+
+        Params:
+            tree = tree to remove the value from
+            node = node currently being inspected for the removal (as we go down
+                     the tree, this parameter is being updated).
+            parent = parent of the node (needed when merging the neighbour nodes)
+            element = element to remove
+            rebalance_parent = indicator if while traversing back to the root, the
+                               parent node should be rebalanced
+
+         Returns: true if the current node needs to be rebalanced
+
+    ******************************************************************************/
+
+    private bool deleteFromNode ( BTreeNode* node, BTreeNode* parent,
```

Comment:
Ok, I've looked into this and I think these methods should stay outside the structure. Reason is that it looks much more natural to look into the implementation, knowing these methods are using recursion to do work, when you're explicitly passing the node on which they are going to work.